src/main/c/NativePtr.cpp

```diff
@@ -0,0 +1,367 @@
+#include <cstdio>
+#include <memory>
+#include <mutex>
+#include <string>
+#include <vector>
+#include <jni.h>
+
+namespace hail {
+
+// JNI interface functions corresponding to Scala NativeBase and NativePtr
+
+#define NATIVEFUNC(typ) \
+  extern "C" JNIEXPORT typ JNICALL
+
+namespace {
+
+void check_assumptions();
+
+class NativePtrInfo {
+public:
+  jclass class_ref_;
+  jfieldID addrA_id_;
+  jfieldID addrB_id_;
+  
+public:
+  NativePtrInfo(JNIEnv* env, int line) {
+    auto cl = env->FindClass("is/hail/nativecode/NativeBase");
+    class_ref_ = (jclass)env->NewGlobalRef(cl);
+    env->DeleteLocalRef(cl);
+    addrA_id_ = env->GetFieldID(class_ref_, "addrA", "J");
+    addrB_id_ = env->GetFieldID(class_ref_, "addrB", "J");
+    check_assumptions();
+  }
+};
+
+// WARNING: I observe that when this gets loaded as a shared library on Linux,
+// we see two distinct NativePtrInfo objects at different addresses.  That is
+// extremely weird, but by putting all the initialization into the constructor
+// we make sure that both get correctly initialized.  But that behavior might
+// cause trouble in other code, so we need to watch out for it.
+
+// We could in theory do the initialization from a JNI_OnLoad() function, but
+// don't want to take time experimenting with that now.
+
+static NativePtrInfo* get_info(JNIEnv* env, int line) {
+  static NativePtrInfo the_info(env, line);
+  return &the_info;
+}
+
+class NativePtrInfo;
+
+// We use this class for moving between a genuine std::shared_ptr<T>,
+// and the Scala NativeBase addrA, addrB.  Sometimes we have a temporary
+// TwoAddrs which we temporarily view as NativeObjPtr; and sometimes we
+// have a genuine NativeObjPtr which we temporarily view as TwoAddrs.
+
+class TwoAddrs {
+public:
+  long addrs_[2];
+
+public:
+  TwoAddrs(JNIEnv* env, jobject obj, NativePtrInfo* info) {
+    addrs_[0] = env->GetLongField(obj, info->addrA_id_);
+    addrs_[1] = env->GetLongField(obj, info->addrB_id_);
+  }
+  
+  TwoAddrs(long addrA, long addrB) {
+    addrs_[0] = addrA;
+    addrs_[1] = addrB;
+  }
+  
+  TwoAddrs(NativeObjPtr&& b) {
+    addrs_[0] = 0;
+    addrs_[1] = 0;
+    this->as_NativeObjPtr() = std::move(b);
+  }
+  
+  TwoAddrs(const NativeObjPtr& b) {
+    addrs_[0] = 0;
+    addrs_[1] = 0;
+    this->as_NativeObjPtr() = b;
+  }
+
+  NativeObjPtr& as_NativeObjPtr() {
+    return *reinterpret_cast<NativeObjPtr*>(this);
+  }
+  
+  void copy_to_scala(JNIEnv* env, jobject obj, NativePtrInfo* info) {
+    env->SetLongField(obj, info->addrA_id_, addrs_[0]);
+    env->SetLongField(obj, info->addrB_id_, addrs_[1]);
+  }
+  
+  long get_addrA() const { return addrs_[0]; }
+  long get_addrB() const { return addrs_[1]; }
+  void set_addrA(long v) { addrs_[0] = v; }
+  void set_addrB(long v) { addrs_[1] = v; }
+};
+
+void check_assumptions() {
+  // Check that std::shared_ptr matches our assumptions
+  auto ptr = std::make_shared<NativeObj>();
+  TwoAddrs* buf = reinterpret_cast<TwoAddrs*>(&ptr);
```

Comment:
So I'm not a huge fan this and the -fno-strict-aliasing.  I would much prefer to have our own lightweight non-threadsafe reference-counted pointer that just exposes the reference count pointer.  It isn't that much code, it would find uses elsewhere, it would make all of this clearer and, maybe most of all, be safer.  Part of it is also that I'm generally less sanguine than you about the performance of shared_ptr, in part because of the refcount overhead, but also because the forced thread safety that I don't normally want.